lib/engine/game/g_18_ka/step/route.rb

```diff
@@ -0,0 +1,124 @@
+  module Game
+    module G18KA
+      module Step
+        class Route < Engine::Step::Route
+          def actions(entity)
+            return [] if !entity.operator? || @game.route_trains(entity).empty? || !@game.can_run_route?(entity)
+
+            @pullman_train_assignments ||= {}
+            actions = ACTIONS.dup
+            actions << 'choose' if !pullman_choices(entity).empty? && !train_choices(entity).empty?
+            actions
+          end
+
+          def attach_pullman(train, pullman)
+            @pullman_train_assignments[pullman] = train
+            city_distance = train_city_distance(train)
+            town_distance = train_town_distance(train)
+            pullman_size = pullman_distance(pullman)
+            train.name = "#{city_distance}+#{town_distance + pullman_size}"
+            train.distance = [
+              {
+                'nodes' => ['town'],
+                'pay' => town_distance + pullman_size,
+                'visit' => town_distance + pullman_size,
+              },
+              {
+                'nodes' => %w[city offboard town],
+                'pay' => city_distance,
+                'visit' => city_distance,
+              },
+            ]
+          end
+
+          def choice_name
+            'Attach a pullman to a train?'
+          end
+
+          def choices
+            choices = {}
+            pullman_choices(current_entity).each_with_index do |pullman, p_index|
+              train_choices(current_entity).each_with_index do |train, t_index|
+                choices["#{t_index}-#{p_index}"] = "Attach #{pullman.name} pullman to #{train.name} train"
+              end
+            end
+            choices
+          end
+
+          def detach_pullmans
+            @pullman_train_assignments.each do |pullman, train|
+              city_distance = train_city_distance(train)
+              town_distance = train_town_distance(train)
+              pullman_size = pullman_distance(pullman)
+              if town_distance == pullman_size
+                train.name = city_distance.to_s
+                train.distance = city_distance
+              else
+                train.name = "#{city_distance}+#{town_distance - pullman_size}"
+                train.distance = [
+                  {
+                    'nodes' => ['town'],
+                    'pay' => town_distance - pullman_size,
+                    'visit' => town_distance - pullman_size,
+                  },
+                  {
+                    'nodes' => %w[city offboard town],
+                    'pay' => city_distance,
+                    'visit' => city_distance,
+                  },
+                ]
+              end
+            end
+            @pullman_train_assignments = {}
+          end
+
+          def train_choices(entity)
+            # Diesels don't get pullmans. That'd be silly.
+            Array(@game.route_trains(entity).reject { |t| @game.pullman_train?(t) || t.variant['name'] == 'D' })
```

Comment:
why are you wrapping this in array?

Reply:
I had `pullman_choices` using `find` originally, Array is no longer needed